test/modellib/ScanModel.cpp

```diff
@@ -0,0 +1,221 @@
+      %3 = "onnx.Identity"(%2) : (tensor<%Bx%Ixf32>) -> tensor<%Bx%Ixf32>
+      "onnx.Return"(%2, %3) : (tensor<%Bx%Ixf32>, tensor<%Bx%Ixf32>) -> ()
+    }) {num_scan_inputs = 1 : si64} :
+        (tensor<%Bx%Ixf32>, tensor<%Bx%Sx%Ixf32>)
+        -> (tensor<%Bx%Ixf32>, tensor<%Bx%Sx%Ixf32>)
+    "func.return"(%1#0, %1#1) : (tensor<%Bx%Ixf32>, tensor<%Bx%Sx%Ixf32>) -> ()
+  }
+  "onnx.EntryPoint"() {func = @main_graph, numInputs = 2 : i32, numOutputs = 2 : i32, signature = "[    ]"} : () -> ()
+})";
+
+std::string testScanIdentityAdd9 = R"(
+module {
+  func @main_graph(%arg0: tensor<%Ixf32>, %arg1: tensor<%Sx%Ixf32>) ->
+                   (tensor<%Ixf32>, tensor<%Sx%Ixf32>) {
+    %1:2 = "onnx.Scan"(%arg0, %arg1) ({
+    ^bb0(%body_arg0: tensor<%Ixf32>, %body_arg1: tensor<%Ixf32>):
+      %2 = "onnx.Add"(%body_arg0, %body_arg1) :
+           (tensor<%Ixf32>, tensor<%Ixf32>) -> tensor<%Ixf32>
+      %3 = "onnx.Identity"(%2) : (tensor<%Ixf32>) -> tensor<%Ixf32>
+      "onnx.Return"(%2, %3) : (tensor<%Ixf32>, tensor<%Ixf32>) -> ()
+    }) {num_scan_inputs = 1 : si64} :
+        (tensor<%Ixf32>, tensor<%Sx%Ixf32>)
+        -> (tensor<%Ixf32>, tensor<%Sx%Ixf32>)
+    "func.return"(%1#0, %1#1) : (tensor<%Ixf32>, tensor<%Sx%Ixf32>) -> ()
+  }
+  "onnx.EntryPoint"() {func = @main_graph, numInputs = 2 : i32, numOutputs = 2 : i32, signature = "[    ]"} : () -> ()
+})";
+
+
+ScanLibBuilder::ScanLibBuilder(const std::string &modelName,
+    const int /*batch=*/B, const int /*seq=*/S, const int /*inner-dim=*/I)
+    : ModelLibBuilder(modelName), B(B), S(S), I(I) {}
+
+bool ScanLibBuilder::build() {
+  initialShape = {B, I};
+  xShape = {B, S, I};
+
+  moduleIR = std::regex_replace(
+      testScanIdentityAdd9, std::regex("%B"), std::to_string(B));
+  moduleIR = std::regex_replace(moduleIR, std::regex("%S"), std::to_string(S));
+  moduleIR = std::regex_replace(moduleIR, std::regex("%I"), std::to_string(I));
+  return true;
+}
+
+bool ScanLibBuilder::compileAndLoad() {
+  OwningOpRef<ModuleOp> moduleOp =
+      mlir::parseSourceString<ModuleOp>(moduleIR, &ctx);
+  OwningOpRef<ModuleOp> module(std::move(moduleOp));
+  if (compileModule(module, ctx, sharedLibBaseName, onnx_mlir::EmitLib) != 0)
+    return false;
+  exec = new ExecutionSession(
+      onnx_mlir::getTargetFilename(sharedLibBaseName, onnx_mlir::EmitLib));
+  return exec != nullptr;
+}
+
+bool ScanLibBuilder::compileAndLoad(const onnx_mlir::CompilerOptionList &list) {
+  if (setCompilerOptions(list) != 0)
+    return false;
+  return compileAndLoad();
+}
+
+bool ScanLibBuilder::prepareInputs() {
+  constexpr int num = 2;
+  OMTensor **list = (OMTensor **)malloc(num * sizeof(OMTensor *));
+  if (!list)
+    return false;
+  list[0] =
+      omTensorCreateWithRandomData<float>(llvm::makeArrayRef(initialShape));
+  list[1] = omTensorCreateWithRandomData<float>(llvm::makeArrayRef(xShape));
+  // Compute reference. Scan with onnx.Add
+  for (int64_t b = 0; b < B; ++b) {
```

Comment:
Why this code here and not in the prepareInputs with range. Also, you could use a single version of the code. I would suggest to define
```C++
template <typename T>
OMTensor *omTensorCreateWithRandomData(
    const std::vector<int64_t> &dataSizes, T lbound = -1.0, T ubound = 1.0);

```
in OMTensorHelper.

I would suggest to define in that file a static def of the default values,
e.g.
```C++
const static float omDefaultRangeBound = 1.0;
```
and then you could use that value there, namely define prepareInputs(float range = omDefaultRangeBound);

If you really wanted, one could even define a default for float and another default for int... but really 0 representation is the same for float and ints...

Reply:
I removed the duplicated part. Thank you!